source/http_connection.c

```diff
@@ -20,6 +20,116 @@
+    (void)env;
+    (void)finalize_hint;
+
+    struct http_proxy_options_binding *binding = finalize_data;
+
+    aws_string_destroy(binding->host_name);
+    aws_string_destroy(binding->auth_username);
+    aws_string_destroy(binding->auth_password);
+
+    aws_mem_release(binding->allocator, binding);
+}
+napi_value aws_napi_http_proxy_options_new(napi_env env, napi_callback_info info) {
+
+    napi_value node_args[6];
+    size_t num_args = AWS_ARRAY_SIZE(node_args);
+
+    napi_value *arg = &node_args[0];
+    if (napi_get_cb_info(env, info, &num_args, node_args, NULL, NULL)) {
+        napi_throw_error(env, NULL, "Failed to retrieve callback information");
+        return NULL;
+    }
+    if (num_args != AWS_ARRAY_SIZE(node_args)) {
+        napi_throw_error(env, NULL, "io_tls_connection_options_new requires exactly 3 arguments");
+        return NULL;
+    }
+
+    napi_value node_external = NULL; /* return value, external that wraps the aws_tls_connection_options */
+
+    struct aws_allocator *allocator = aws_default_allocator();
+    struct http_proxy_options_binding *binding =
+        aws_mem_calloc(allocator, 1, sizeof(struct http_proxy_options_binding));
+    AWS_FATAL_ASSERT(binding && "Failed to allocate new http_proxy_options_binding");
+    binding->allocator = allocator;
+
+    napi_value node_host_name = *arg++;
+    binding->host_name = aws_string_new_from_napi(env, node_host_name);
+    if (!binding->host_name) {
+        AWS_NAPI_ENSURE(env, napi_throw_type_error(env, NULL, "Unable to convert host_name to string"));
+        goto cleanup;
+    }
+
+    napi_value node_port = *arg++;
+    uint32_t port;
+    AWS_NAPI_CALL(env, napi_get_value_uint32(env, node_port, &port), {
+        napi_throw_type_error(env, NULL, "port must be a number");
+        goto cleanup;
+    });
+    binding->native.port = port;
+
+    napi_value node_auth_method = *arg++;
+    if (!aws_napi_is_null_or_undefined(env, node_auth_method)) {
+        uint32_t auth_method = 0;
+        AWS_NAPI_CALL(env, napi_get_value_uint32(env, node_port, &auth_method), {
+            napi_throw_type_error(env, NULL, "auth_method must be a number");
+            goto cleanup;
+        });
+        binding->native.auth_type = auth_method;
+    }
+
+    napi_value node_username = *arg++;
+    if (!aws_napi_is_null_or_undefined(env, node_username)) {
+        binding->auth_username = aws_string_new_from_napi(env, node_username);
+        if (!binding->auth_username) {
+            AWS_NAPI_ENSURE(env, napi_throw_type_error(env, NULL, "Unable to convert auth_username to string"));
+            goto cleanup;
+        }
+    }
+
+    napi_value node_password = *arg++;
+    if (!aws_napi_is_null_or_undefined(env, node_password)) {
+        binding->auth_password = aws_string_new_from_napi(env, node_password);
+        if (!binding->auth_password) {
+            AWS_NAPI_ENSURE(env, napi_throw_type_error(env, NULL, "Unable to convert auth_password to string"));
```

Comment:
Why is this type error ENSURE'd but none of the other ones are? I don't think it matters if the throw fails, nothing you can do about it, and the next call will clean it up.